### ISSUE TITLE
Can't write to an existing file

### ISSUE THREAD

**Issue by VBaratham** [NeurodataWithoutBorders/pynwb#662]
## 1) Bug

pynwb throws an error when you append to an existing file that contains an electrode_group

### Steps to Reproduce
```python
from datetime import datetime
import numpy as np
from pynwb import NWBFile, NWBHDF5IO, TimeSeries
from pynwb.ecephys import ElectricalSeries, LFP

FILENAME = 'test_append.nwb'

if __name__ == '__main__':
    nwb = NWBFile(source='', session_description='', identifier='', session_start_time=datetime.now())
    proc_mod = nwb.create_processing_module(name='test_proc_mod', source='', description='')
    proc_inter = LFP(source='', name='test_proc_dset')
    proc_mod.add_data_interface(proc_inter)
    device = nwb.create_device(name='test_device', source='')
    e_group = nwb.create_electrode_group(
        name='test_electrode_group',
        source='',
        description='',
        location='',
        device=device
    )

    with NWBHDF5IO(FILENAME, mode='w') as io:
        io.write(nwb)

    with NWBHDF5IO(FILENAME, mode='a') as io:
        nwb = io.read()
        ts2 = TimeSeries(name='timeseries2', source='source2', data=[4,5,6],
                unit='na1', rate=1.0, starting_time=0.0)
        nwb.add_acquisition(ts2)
        io.write(nwb)
```
The error is:
```
Traceback (most recent call last):
  File "scripts/test_nwb_append.py", line 58, in <module>
    io.write(nwb)
  File "/home/vyassa/anaconda2/envs/mars/lib/python3.5/site-packages/pynwb/form/utils.py", line 337, in func_call
    return func(self, **parsed['args'])
  File "/home/vyassa/anaconda2/envs/mars/lib/python3.5/site-packages/pynwb/form/backends/hdf5/h5tools.py", line 186, in write
    call_docval_func(super(HDF5IO, self).write, kwargs)
  File "/home/vyassa/anaconda2/envs/mars/lib/python3.5/site-packages/pynwb/form/utils.py", line 235, in call_docval_func
    return func(*fargs, **fkwargs)
  File "/home/vyassa/anaconda2/envs/mars/lib/python3.5/site-packages/pynwb/form/utils.py", line 337, in func_call
    return func(self, **parsed['args'])
  File "/home/vyassa/anaconda2/envs/mars/lib/python3.5/site-packages/pynwb/form/backends/io.py", line 40, in write
    self.write_builder(f_builder, **kwargs)
  File "/home/vyassa/anaconda2/envs/mars/lib/python3.5/site-packages/pynwb/form/utils.py", line 337, in func_call
    return func(self, **parsed['args'])
  File "/home/vyassa/anaconda2/envs/mars/lib/python3.5/site-packages/pynwb/form/backends/hdf5/h5tools.py", line 367, in write_builder
    self.write_group(self.__file, gbldr)
  File "/home/vyassa/anaconda2/envs/mars/lib/python3.5/site-packages/pynwb/form/utils.py", line 337, in func_call
    return func(self, **parsed['args'])
  File "/home/vyassa/anaconda2/envs/mars/lib/python3.5/site-packages/pynwb/form/backends/hdf5/h5tools.py", line 504, in write_group
    self.write_group(group, sub_builder)
  File "/home/vyassa/anaconda2/envs/mars/lib/python3.5/site-packages/pynwb/form/utils.py", line 337, in func_call
    return func(self, **parsed['args'])
  File "/home/vyassa/anaconda2/envs/mars/lib/python3.5/site-packages/pynwb/form/backends/hdf5/h5tools.py", line 504, in write_group
    self.write_group(group, sub_builder)
  File "/home/vyassa/anaconda2/envs/mars/lib/python3.5/site-packages/pynwb/form/utils.py", line 337, in func_call
    return func(self, **parsed['args'])
  File "/home/vyassa/anaconda2/envs/mars/lib/python3.5/site-packages/pynwb/form/backends/hdf5/h5tools.py", line 514, in write_group
    self.write_link(group, sub_builder)
  File "/home/vyassa/anaconda2/envs/mars/lib/python3.5/site-packages/pynwb/form/utils.py", line 337, in func_call
    return func(self, **parsed['args'])
  File "/home/vyassa/anaconda2/envs/mars/lib/python3.5/site-packages/pynwb/form/backends/hdf5/h5tools.py", line 548, in write_link
    parent[name] = link_obj
  File "/home/vyassa/anaconda2/envs/mars/lib/python3.5/site-packages/h5py/_hl/group.py", line 291, in __setitem__
    lcpl=lcpl, lapl=self._lapl)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5l.pyx", line 145, in h5py.h5l.LinkProxy.create_soft
RuntimeError: Unable to create link (name already exists)
```
### Environment
- **Python Executable:** Conda
- **Python Version:** Python 3.5.1
- **Operating System:** Linux (Ubuntu 18.04)
- **Pynwb Version:** pynwb 0.5.1

## Checklist

- [x] Have you ensured the feature or change was not already [reported](https://github.com/NeurodataWithoutBorders/pynwb/issues/) ?
- [x] Have you included a brief and descriptive title?
- [x] Have you included a clear description of the problem you are trying to solve?
- [x] Have you included a minimal code snippet that reproduces the issue you are encountering?

**Comment by ageorgou** [comment]
I've come across this as well. In my case, it was solved by updating the NWBFile after each write. I played around with this example, and noticed two things:

1. The first write has to be done in `w` mode, not `a`. This is the first problem here as well, since the error is caused at the first write (line 38), not the one marked with `# fails`.

2. Assuming you update the file after the first write (see below), there is a problem with writing the electrode group at the second write:

```
Traceback (most recent call last):
  File "test_github2.py", line 41, in <module>
    io.write(nwb) # fails
  File "/Users/ageorgou/anaconda/envs/nwb/lib/python3.5/site-packages/pynwb/form/utils.py", line 337, in func_call
    return func(self, **parsed['args'])
  File "/Users/ageorgou/anaconda/envs/nwb/lib/python3.5/site-packages/pynwb/form/backends/hdf5/h5tools.py", line 186, in write
    call_docval_func(super(HDF5IO, self).write, kwargs)
  File "/Users/ageorgou/anaconda/envs/nwb/lib/python3.5/site-packages/pynwb/form/utils.py", line 235, in call_docval_func
    return func(*fargs, **fkwargs)
  File "/Users/ageorgou/anaconda/envs/nwb/lib/python3.5/site-packages/pynwb/form/utils.py", line 337, in func_call
    return func(self, **parsed['args'])
  File "/Users/ageorgou/anaconda/envs/nwb/lib/python3.5/site-packages/pynwb/form/backends/io.py", line 40, in write
    self.write_builder(f_builder, **kwargs)
  File "/Users/ageorgou/anaconda/envs/nwb/lib/python3.5/site-packages/pynwb/form/utils.py", line 337, in func_call
    return func(self, **parsed['args'])
  File "/Users/ageorgou/anaconda/envs/nwb/lib/python3.5/site-packages/pynwb/form/backends/hdf5/h5tools.py", line 367, in write_builder
    self.write_group(self.__file, gbldr)
  File "/Users/ageorgou/anaconda/envs/nwb/lib/python3.5/site-packages/pynwb/form/utils.py", line 337, in func_call
    return func(self, **parsed['args'])
  File "/Users/ageorgou/anaconda/envs/nwb/lib/python3.5/site-packages/pynwb/form/backends/hdf5/h5tools.py", line 504, in write_group
    self.write_group(group, sub_builder)
  File "/Users/ageorgou/anaconda/envs/nwb/lib/python3.5/site-packages/pynwb/form/utils.py", line 337, in func_call
    return func(self, **parsed['args'])
  File "/Users/ageorgou/anaconda/envs/nwb/lib/python3.5/site-packages/pynwb/form/backends/hdf5/h5tools.py", line 504, in write_group
    self.write_group(group, sub_builder)
  File "/Users/ageorgou/anaconda/envs/nwb/lib/python3.5/site-packages/pynwb/form/utils.py", line 337, in func_call
    return func(self, **parsed['args'])
  File "/Users/ageorgou/anaconda/envs/nwb/lib/python3.5/site-packages/pynwb/form/backends/hdf5/h5tools.py", line 514, in write_group
    self.write_link(group, sub_builder)
  File "/Users/ageorgou/anaconda/envs/nwb/lib/python3.5/site-packages/pynwb/form/utils.py", line 337, in func_call
    return func(self, **parsed['args'])
  File "/Users/ageorgou/anaconda/envs/nwb/lib/python3.5/site-packages/pynwb/form/backends/hdf5/h5tools.py", line 548, in write_link
    parent[name] = link_obj
  File "/Users/ageorgou/anaconda/envs/nwb/lib/python3.5/site-packages/h5py/_hl/group.py", line 291, in __setitem__
    lcpl=lcpl, lapl=self._lapl)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5l.pyx", line 145, in h5py.h5l.LinkProxy.create_soft
RuntimeError: Unable to create link (name already exists)
```
Note that the error message and the last line of the trace are different (the line number is off by one compared to the original example due to adding a line to update the NWBFile, as below).

If you only add the device, then the second write works fine. So the following code should run without errors:

```python
from datetime import datetime

import numpy as np

from pynwb import NWBFile, NWBHDF5IO
from pynwb.ecephys import ElectricalSeries, LFP


FILENAME = 'test_append.nwb'

if __name__ == '__main__':
    nwb = NWBFile(source='', session_description='', identifier='', session_start_time=datetime.now())
    proc_mod = nwb.create_processing_module(name='test_proc_mod', source='', description='')
    proc_inter = LFP(source='', name='test_proc_dset')
    proc_mod.add_data_interface(proc_inter)
    device = nwb.create_device(name='test_device', source='')
    # Commenting out electrode group
    # e_group = nwb.create_electrode_group(
    #     name='test_electrode_group',
    #     source='',
    #     description='',
    #     location='',
    #     device=device
    # )
    # nwb.add_electrode(0, x=0.0, y=0.0, z=0.0, imp=np.nan, location='', filtering='', description='', group=e_group)
    # electrodes = nwb.create_electrode_table_region(region=[0], description='')
    # e_series = ElectricalSeries(
    #     name='test_device',
    #     source='',
    #     electrodes=electrodes,
    #     data=np.ones(shape=(100,)),
    #     starting_time=0.0,
    #     rate=10000.0,
    # )
    # proc_inter.add_electrical_series(e_series)

    # This throws an error with 'a' mode  (even if file already exists)
    with NWBHDF5IO(FILENAME, mode='a') as io:
        io.write(nwb)
    with NWBHDF5IO(FILENAME, mode='a') as io:
        # We need to sync the NWBFile with the disk file contents
        # (this has to be done in this context, not the previous one)
        nwb = io.read()
        io.write(nwb) # now succeeds
```

I'm not sure whether the requirement to update the NWBFile when working in `'a'` mode is intended behaviour or not (the code sample in #615 seems to suggest that's the way to do it).

Sorry to hijack your issue, @VBaratham. Hope this helps.

**Comment by VBaratham** [comment]
Thank you @ageorgou this helps a lot. I'll talk to @ajtritt today and see if we can do anything about the electrode group

**Comment by ajtritt** [comment]
The error here was caused by the link created from the ElectrodeGroup to the Device. When this link was read, it wasn't being tracked as already being written. So, on the second write, it was trying to create it again, hence the error `Unable to create link (name already exists)`. I fixed this tracking and this issue should be resolved when #665 gets merged.
